### ISSUE TITLE
ansible_ssh_user not being picked up

### ISSUE THREAD

**Issue by jtktam** [ansible/ansible#11132]
updated the latest version and ansible is not picking up the ansible_ssh_user in my group_vars, if i specify it on the command line it works fine.

---
# development.yml

```

---
[lbservers]
dev-lb1 ansible_ssh_host=x.x.x.x

[webservers]
dev-www1 ansible_ssh_host=y.y.y.y

[dbservers]

[allservers:children]
lbservers
webservers
dbservers

[devservers:children]
allservers
```

---
# group_vars/allservers.yml

```

---
ansible_ssh_user: root

----------
# webservers.yml

---
- hosts: webservers
  roles:
    - php5
```

---

```
ansible-playbook -i development.yml lbservers.yml

PLAY: ***************************************************************************

TASK [setup] ********************************************************************
fatal: [dev-www1]: FAILED! => {"msg": "ERROR! SSH encountered an unknown error. The output was:\n", "failed": true}
```

---

```
ansible-playbook -i development.yml lbservers.yml --user=root

PLAY: ***************************************************************************

TASK [setup] ********************************************************************
ok: [dev-www1]

PLAY: ***************************************************************************
```

**Comment by jtktam** [comment]
Actually looks like the entire group_vars/allservers.yml is not being picked up, there are other variables in there and they are returning as undefined in the scripts

---

ansible-playbook -i development.yml lbservers.yml --user=root

PLAY: ***************************************************************************

TASK [setup] ********************************************************************
ok: [dev-www1]

PLAY: ***************************************************************************

TASK [setup] ********************************************************************
ok: [dev-lb1]

TASK [common : Copy firewall rules] *********************************************
fatal: [dev-lb1]: FAILED! => {"msg": "AnsibleUndefinedVariable: ERROR! Unable to look up a name or access an attribute in template string. Make sure your variable name does not contain invalid characters like '-'.", "failed": true, "changed": false}

**Comment by bcoca** [comment]
possibly related/overlapped with #11130

**Comment by jtktam** [comment]
# group_vars/allservers.yml

```
ansible_ssh_user: root

public_ip: "{{ ansible_eth0.ipv4.address }}"
private_ip: "{{ ansible_eth1.ipv4.address }}"

public_iface: "{{ ansible_eth0.device }}"
private_iface: "{{ ansible_eth1.device }}"

public_httpd_port: 80
private_httpd_port: 80
```

---

# roles/common/templates/iptables.sav.j2

```
*filter
:INPUT DROP [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [3:344]

{% if inventory_hostname in groups['webservers'] %}
-A INPUT -i {{ private_iface }} -p tcp -m tcp --dport {{ private_httpd_port }} -j ACCEPT
{% endif %}

{% if inventory_hostname in groups['lbservers'] %}
-A INPUT -i {{ public_iface }} -p tcp -m tcp --dport {{ public_httpd_port }} -j ACCEPT
{% endif %}
```

---

it's failing on the if / endif blocks, it can't find private_httpd_port, which is part of the allservers.yml file

so i can conclude that the entire file is not being picked up as group vars

if you need anything else to reproduce this issue let me know

**Comment by legleux** [comment]
I seem to have the same problem here. My inventory file is in the same dir as my host_var and group_vars and they are not being used by ansible-playbook.

**Comment by bcoca** [comment]
do you have --- in the middle of your yaml files?

**Comment by jtktam** [comment]
no, and the exact same playbook / groupvars / etc works last week, i did a github update and now it broken :(

just revert backed to v1.9.2-0.1.rc1 and confirm my playbook / group_vars are working again, something between that version and the latest is causing some issues

**Comment by legleux** [comment]
Previously, I started using ansible 2.0.0 about 3 weeks ago so I was working with 2.0 for awhile. Then I did a pull on monday and no more variables.

**Comment by bcoca** [comment]
trying to clarify but probably will add to the confusion. current devel has been 2.0 since 1.9 was released, but Monday we changed the main ansible engine/core to the v2 version that has been in development in parallel, we expected a few bugs and this seems to be one of them, we will be fixing this soon, but we also have ansible fest NYC tomorrow, so it might take us a bit longer than normal.

**Comment by sivel** [comment]
I've done a little digging, and have something working.

I do see that `group_vars` in general are working, just not for the special connection related inventory variables such as `ansible_ssh_user`.

This is largely because `group_vars` have been moved out of inventory, and into the `VariableManager`.  `ConnectionInfo` does not have access to the `VariableManager`.  When `ConnectionInfo.set_task_and_host_override` is called, it is only passed a `Host` instance, which no longer includes the variables set in `group_vars`.

Not that this is the "correct" solution as their are several ways to approach this, but at the moment, this appears to be working:

``` diff
diff --git a/lib/ansible/executor/connection_info.py b/lib/ansible/executor/connection_info.py
index 424ac06..db3d41b 100644
--- a/lib/ansible/executor/connection_info.py
+++ b/lib/ansible/executor/connection_info.py
@@ -177,7 +177,7 @@ class ConnectionInformation:
             else:
                 setattr(self, field, value)

-    def set_task_and_host_override(self, task, host):
+    def set_task_and_host_override(self, variables):
         '''
         Sets attributes from the task if they are set, which will override
         those from the play.
@@ -189,14 +189,11 @@ class ConnectionInformation:
         # loop through a subset of attributes on the task object and set
         # connection fields based on their values
         for attr in ('connection', 'remote_user', 'become', 'become_user', 'become_pass', 'become_method', 'environment', 'no_log'):
-            if hasattr(task, attr):
-                attr_val = getattr(task, attr)
-                if attr_val is not None:
-                    setattr(new_info, attr, attr_val)
+            if variables.get(attr):
+                setattr(new_info, attr, variables.get(attr))

         # finally, use the MAGIC_VARIABLE_MAPPING dictionary to update this
         # connection info object with 'magic' variables from inventory
-        variables = host.get_vars()
         for (attr, variable_names) in MAGIC_VARIABLE_MAPPING.iteritems():
             for variable_name in variable_names:
                 if variable_name in variables:
diff --git a/lib/ansible/executor/process/worker.py b/lib/ansible/executor/process/worker.py
index e1488eb..e924e92 100644
--- a/lib/ansible/executor/process/worker.py
+++ b/lib/ansible/executor/process/worker.py
@@ -51,12 +51,13 @@ class WorkerProcess(multiprocessing.Process):
     for reading later.
     '''

-    def __init__(self, tqm, main_q, rslt_q, loader):
+    def __init__(self, tqm, main_q, rslt_q, loader, variable_manager):

         # takes a task queue manager as the sole param:
         self._main_q = main_q
         self._rslt_q = rslt_q
         self._loader = loader
+        self._variable_manager = variable_manager

         # dupe stdin, if we have one
         self._new_stdin = sys.stdin
@@ -111,7 +112,8 @@ class WorkerProcess(multiprocessing.Process):
                     # apply the given task's information to the connection info,
                     # which may override some fields already set by the play or
                     # the options specified on the command line
-                    new_connection_info = connection_info.set_task_and_host_override(task=task, host=host)
+                    variables = self._variable_manager.get_vars(self._loader, host=host, task=task)
+                    new_connection_info = connection_info.set_task_and_host_override(variables)

                     # execute the task and build a TaskResult from the result
                     debug("running TaskExecutor() for %s/%s" % (host, task))
diff --git a/lib/ansible/executor/task_queue_manager.py b/lib/ansible/executor/task_queue_manager.py
index a875c31..9838a26 100644
--- a/lib/ansible/executor/task_queue_manager.py
+++ b/lib/ansible/executor/task_queue_manager.py
@@ -86,7 +86,7 @@ class TaskQueueManager:
             main_q = multiprocessing.Queue()
             rslt_q = multiprocessing.Queue()

-            prc = WorkerProcess(self, main_q, rslt_q, loader)
+            prc = WorkerProcess(self, main_q, rslt_q, loader, variable_manager)
             prc.start()

             self._workers.append((prc, main_q, rslt_q))
```

**Comment by sivel** [comment]
I can create a PR with the above after talking it out with some of the core deva in more depth.

**Comment by bcoca** [comment]
need to check with james, but I believe the intention was for those vars to be attached to the host and override from there.

**Comment by jimi-c** [comment]
The issue here looks to be very simple: the group/host var loading methods only look for the name of the host/group specifically, and don't try to find .yml/.yaml files too. Combined with a modification of the fix in #11218 I believe I have this resolved and will push up a fix soon.

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
